contracts/schemes/SchemeFactory.sol

```diff
@@ -0,0 +1,172 @@
+    event NewSchemeProposal(
+        address indexed _avatar,
+        bytes32 indexed _proposalId,
+        address indexed _intVoteInterface,
+        string _schemeName,
+        bytes _schemeData,
+        uint64[3] _packageVersion,
+        bytes4 _permissions,
+        address _schemeToReplace,
+        string _descriptionHash
+    );
+
+    event ProposalExecuted(address indexed _avatar, bytes32 indexed _proposalId, int256 _param);
+    event ProposalDeleted(address indexed _avatar, bytes32 indexed _proposalId);
+
+    // a proposal to add or remove a scheme to/from the an organization
+    struct Proposal {
+        string schemeName;
+        bytes schemeData;
+        uint64[3] packageVersion;
+        address schemeToReplace;
+        bytes4 permissions;
+    }
+
+    mapping(bytes32=>Proposal) public proposals;
+
+    IntVoteInterface public votingMachine;
+    bytes32 public voteParams;
+    Avatar public avatar;
+    DAOFactory public daoFactory;
+
+    /**
+     * @dev initialize
+     * @param _avatar the avatar this scheme referring to.
+     * @param _votingMachine the voting machines address to
+     * @param _voteParams voting machine parameters to register scheme.
+     */
+    function initialize(
+        Avatar _avatar,
+        IntVoteInterface _votingMachine,
+        bytes32 _voteParams,
+        DAOFactory _daoFactory
+    )
+    external
+    initializer
+    {
+        require(_avatar != Avatar(0), "avatar cannot be zero");
+        avatar = _avatar;
+        votingMachine = _votingMachine;
+        voteParams = _voteParams;
+        daoFactory = _daoFactory;
+    }
+
+    /**
+    * @dev execution of proposals, can only be called by the voting machine in which the vote is held.
+    * @param _proposalId the ID of the voting in the voting machine
+    * @param _decision the voting result, 1 yes and 2 is no.
+    * @return true (if function did not revert)
+    */
+    function executeProposal(bytes32 _proposalId, int256 _decision)
+    external
+    onlyVotingMachine(_proposalId)
+    returns(bool) {
+        Proposal memory proposal = proposals[_proposalId];
+        delete proposals[_proposalId];
+        emit ProposalDeleted(address(avatar), _proposalId);
+        Controller controller = Controller(avatar.owner());
+        if (_decision == 1) {
+            if (bytes(proposal.schemeName).length > 0) {
+                address scheme = address(daoFactory.createInstance(
+                                        proposal.packageVersion,
+                                        proposal.schemeName,
+                                        address(avatar),
+                                        proposal.schemeData));
+
+                require(controller.registerScheme(scheme, proposal.permissions), "faild to register new scheme");
+            }
+
+
+            if (proposal.schemeToReplace != address(0) && controller.isSchemeRegistered(proposal.schemeToReplace)) {
```

Comment:
controller.isSchemeRegistered is checked already at line 131 ?

Reply:
I'm checking here again because it is possible that meanwhile, as the proposal was in voting phase, the scheme was already removed by another proposal/scheme.

Reply:
It's not a must I guess, but could be a good check to have

Reply:
I am not sure that on this case we want to revert ..though.
I think it is better that the execution will not revert as it might do a problem with the gov process (e.g the last vote which cause execution).

Reply:
please consider not emitting the ProposalExecuted event or emit another event ..as it might confused the upper layers

Reply:
It doesn't cause a revert here though. It just won't run the unregister

Reply:
This if here is actually necessary to prevent a revert if the scheme is not registered anymore for some reason. Without it, the next line will revert (as the unregister will return false)